<h1>ACTIVITY 1 - PALMER PEGUINS DATA ANALYTICS PROJECT </h1>
<hr>
<h3> Analyst : Junelle C Dinglasan </h3>

In [1]:
#import libraries
import numpy as np
import pandas as pd

<h3>Converting Dataset(s) into DataFrame</h3>
<h2>Requirements</h2>
<ul>
    <li>Name of the Dataframe</li>
    <li>File Path</li>
</ul>

In [2]:
penguins = pd.read_csv('datasets\\penguins.csv')

In [3]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

In [4]:
penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      342.000000     342.000000         342.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.459584       1.974793          14.061714   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.225000      15.600000         190.000000  3550.000000   
50%         44.450000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

In [5]:
penguins.describe().T

count         mean         std     min       25%      50%  \
bill_length_mm     342.0    43.921930    5.459584    32.1    39.225    44.45   
bill_depth_mm      342.0    17.151170    1.974793    13.1    15.600    17.30   
flipper_length_mm  342.0   200.915205   14.061714   172.0   190.000   197.00   
body_mass_g        342.0  4201.754386  801.954536  2700.0  3550.000  4050.00   
year               344.0  2008.029070    0.818356  2007.0  2007.000  2008.00   

                      75%     max  
bill_length_mm       48.5    59.6  
bill_depth_mm        18.7    21.5  
flipper_length_mm   213.0   231.0  
body_mass_g        4750.0  6300.0  
year               2009.0  2009.0

<h3>Checking for Null values</h3>

In [6]:
penguins.isnull()

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0      False   False           False          False              False   
1      False   False           False          False              False   
2      False   False           False          False              False   
3      False   False            True           True               True   
4      False   False           False          False              False   
..       ...     ...             ...            ...                ...   
339    False   False           False          False              False   
340    False   False           False          False              False   
341    False   False           False          False              False   
342    False   False           False          False              False   
343    False   False           False          False              False   

     body_mass_g    sex   year  
0          False  False  False  
1          False  False  False  
2          False  False  False  
3           True   True  False  
4          False  False  False  
..           ...    ...    ...  
339        False  False  False  
340        False  False  False  
341        False  False  False  
342        False  False  False  
343        False  False  False  

[344 rows x 8 columns]

In [7]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [8]:
penguins['species']

0         Adelie
1         Adelie
2         Adelie
3         Adelie
4         Adelie
         ...    
339    Chinstrap
340    Chinstrap
341    Chinstrap
342    Chinstrap
343    Chinstrap
Name: species, Length: 344, dtype: object

In [9]:
penguins['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

In [10]:
penguins[penguins['species'] == 'Chinstrap']

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
276  Chinstrap  Dream            46.5           17.9              192.0   
277  Chinstrap  Dream            50.0           19.5              196.0   
278  Chinstrap  Dream            51.3           19.2              193.0   
279  Chinstrap  Dream            45.4           18.7              188.0   
280  Chinstrap  Dream            52.7           19.8              197.0   
..         ...    ...             ...            ...                ...   
339  Chinstrap  Dream            55.8           19.8              207.0   
340  Chinstrap  Dream            43.5           18.1              202.0   
341  Chinstrap  Dream            49.6           18.2              193.0   
342  Chinstrap  Dream            50.8           19.0              210.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
276       3500.0  female  2007  
277       3900.0    male  2007  
278       3650.0    male  2007  
279       3525.0  female  2007  
280       3725.0    male  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[68 rows x 8 columns]

In [11]:
chinstrap_df = penguins[penguins['species'] == 'Chinstrap']

In [12]:
chinstrap_df

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
276  Chinstrap  Dream            46.5           17.9              192.0   
277  Chinstrap  Dream            50.0           19.5              196.0   
278  Chinstrap  Dream            51.3           19.2              193.0   
279  Chinstrap  Dream            45.4           18.7              188.0   
280  Chinstrap  Dream            52.7           19.8              197.0   
..         ...    ...             ...            ...                ...   
339  Chinstrap  Dream            55.8           19.8              207.0   
340  Chinstrap  Dream            43.5           18.1              202.0   
341  Chinstrap  Dream            49.6           18.2              193.0   
342  Chinstrap  Dream            50.8           19.0              210.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
276       3500.0  female  2007  
277       3900.0    male  2007  
278       3650.0    male  2007  
279       3525.0  female  2007  
280       3725.0    male  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[68 rows x 8 columns]

In [13]:
adelie_df = penguins[penguins['species'] == 'Adelie']

In [14]:
adelie_df

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
147  Adelie      Dream            36.6           18.4              184.0   
148  Adelie      Dream            36.0           17.8              195.0   
149  Adelie      Dream            37.8           18.1              193.0   
150  Adelie      Dream            36.0           17.1              187.0   
151  Adelie      Dream            41.5           18.5              201.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
147       3475.0  female  2009  
148       3450.0  female  2009  
149       3750.0    male  2009  
150       3700.0  female  2009  
151       4000.0    male  2009  

[152 rows x 8 columns]

In [15]:
gentoo_df = penguins[penguins['species'] == 'Gentoo']

In [16]:
gentoo_df

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
152  Gentoo  Biscoe            46.1           13.2              211.0   
153  Gentoo  Biscoe            50.0           16.3              230.0   
154  Gentoo  Biscoe            48.7           14.1              210.0   
155  Gentoo  Biscoe            50.0           15.2              218.0   
156  Gentoo  Biscoe            47.6           14.5              215.0   
..      ...     ...             ...            ...                ...   
271  Gentoo  Biscoe             NaN            NaN                NaN   
272  Gentoo  Biscoe            46.8           14.3              215.0   
273  Gentoo  Biscoe            50.4           15.7              222.0   
274  Gentoo  Biscoe            45.2           14.8              212.0   
275  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  year  
152       4500.0  female  2007  
153       5700.0    male  2007  
154       4450.0  female  2007  
155       5700.0    male  2007  
156       5400.0    male  2007  
..           ...     ...   ...  
271          NaN     NaN  2009  
272       4850.0  female  2009  
273       5750.0    male  2009  
274       5200.0  female  2009  
275       5400.0    male  2009  

[124 rows x 8 columns]

In [17]:
penguins[penguins['body_mass_g'] >= 4000]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
7       Adelie  Torgersen            39.2           19.6              195.0   
9       Adelie  Torgersen            42.0           20.2              190.0   
14      Adelie  Torgersen            34.6           21.1              198.0   
17      Adelie  Torgersen            42.5           20.7              197.0   
19      Adelie  Torgersen            46.0           21.5              194.0   
..         ...        ...             ...            ...                ...   
323  Chinstrap      Dream            49.0           19.6              212.0   
329  Chinstrap      Dream            50.7           19.7              203.0   
333  Chinstrap      Dream            49.3           19.9              203.0   
339  Chinstrap      Dream            55.8           19.8              207.0   
342  Chinstrap      Dream            50.8           19.0              210.0   

     body_mass_g   sex  year  
7         4675.0  male  2007  
9         4250.0   NaN  2007  
14        4400.0  male  2007  
17        4500.0  male  2007  
19        4200.0  male  2007  
..           ...   ...   ...  
323       4300.0  male  2009  
329       4050.0  male  2009  
333       4050.0  male  2009  
339       4000.0  male  2009  
342       4100.0  male  2009  

[177 rows x 8 columns]

In [18]:
chinstrap_heavy = penguins[(penguins['species'] == 'Chinstrap') & (penguins['body_mass_g'] >= 4000)]

In [19]:
chinstrap_heavy

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
284  Chinstrap  Dream            46.0           18.9              195.0   
289  Chinstrap  Dream            52.0           18.1              201.0   
291  Chinstrap  Dream            50.5           19.6              201.0   
295  Chinstrap  Dream            49.2           18.2              195.0   
301  Chinstrap  Dream            52.0           19.0              197.0   
305  Chinstrap  Dream            52.8           20.0              205.0   
307  Chinstrap  Dream            54.2           20.8              201.0   
309  Chinstrap  Dream            51.0           18.8              203.0   
313  Chinstrap  Dream            52.0           20.7              210.0   
315  Chinstrap  Dream            53.5           19.9              205.0   
321  Chinstrap  Dream            50.8           18.5              201.0   
323  Chinstrap  Dream            49.0           19.6              212.0   
329  Chinstrap  Dream            50.7           19.7              203.0   
333  Chinstrap  Dream            49.3           19.9              203.0   
339  Chinstrap  Dream            55.8           19.8              207.0   
342  Chinstrap  Dream            50.8           19.0              210.0   

     body_mass_g     sex  year  
284       4150.0  female  2007  
289       4050.0    male  2007  
291       4050.0    male  2007  
295       4400.0    male  2007  
301       4150.0    male  2007  
305       4550.0    male  2008  
307       4300.0    male  2008  
309       4100.0    male  2008  
313       4800.0    male  2008  
315       4500.0    male  2008  
321       4450.0    male  2009  
323       4300.0    male  2009  
329       4050.0    male  2009  
333       4050.0    male  2009  
339       4000.0    male  2009  
342       4100.0    male  2009

In [20]:
gentoo_heavy = penguins[(penguins['species'] == 'Gentoo') & (penguins['body_mass_g'] >= 4000)]

In [21]:
gentoo_heavy

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
152  Gentoo  Biscoe            46.1           13.2              211.0   
153  Gentoo  Biscoe            50.0           16.3              230.0   
154  Gentoo  Biscoe            48.7           14.1              210.0   
155  Gentoo  Biscoe            50.0           15.2              218.0   
156  Gentoo  Biscoe            47.6           14.5              215.0   
..      ...     ...             ...            ...                ...   
270  Gentoo  Biscoe            47.2           13.7              214.0   
272  Gentoo  Biscoe            46.8           14.3              215.0   
273  Gentoo  Biscoe            50.4           15.7              222.0   
274  Gentoo  Biscoe            45.2           14.8              212.0   
275  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  year  
152       4500.0  female  2007  
153       5700.0    male  2007  
154       4450.0  female  2007  
155       5700.0    male  2007  
156       5400.0    male  2007  
..           ...     ...   ...  
270       4925.0  female  2009  
272       4850.0  female  2009  
273       5750.0    male  2009  
274       5200.0  female  2009  
275       5400.0    male  2009  

[122 rows x 8 columns]

In [22]:
penguins['bill_length_mm'].mean()

np.float64(43.9219298245614)

In [23]:
penguins['bill_length_mm'] = penguins['bill_length_mm'].fillna(penguins['bill_length_mm'].mean())

In [24]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [25]:
penguins['bill_depth_mm'].mean()

np.float64(17.151169590643274)

In [26]:
penguins['bill_depth_mm'] = penguins['bill_depth_mm'].fillna(penguins['bill_depth_mm'].mean())

In [27]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [28]:
penguins['flipper_length_mm'].mean()

np.float64(200.91520467836258)

In [29]:
penguins['flipper_length_mm'] = penguins['flipper_length_mm'].fillna(penguins['flipper_length_mm'].mean())

In [30]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     0
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [31]:
penguins['sex'].mode()

0    male
Name: sex, dtype: object

In [32]:
penguins['sex'] = penguins['sex'].fillna(penguins['sex'].mode()[0])

In [33]:
penguins.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          2
sex                  0
year                 0
dtype: int64

In [34]:
penguins['body_mass_g'].mean()

np.float64(4201.754385964912)

In [35]:
penguins['body_mass_g'] = penguins['body_mass_g'].fillna(penguins['body_mass_g'].mean())

In [36]:
penguins.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

<h3>Changing the data types</h3>

In [37]:
penguins['species'] = penguins['species'].astype('category')
penguins['year'] = penguins['year'].astype('category')
penguins['island'] = penguins['island'].astype('category')
penguins['sex'] = penguins['sex'].astype('category')

In [38]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    category
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      344 non-null    float64 
 4   flipper_length_mm  344 non-null    float64 
 5   body_mass_g        344 non-null    float64 
 6   sex                344 non-null    category
 7   year               344 non-null    category
dtypes: category(4), float64(4)
memory usage: 12.7 KB


<h3>Exploratory Data Analytics</h3>

<h3>Question 1</h3>
<h3>Which penguins have a bill length between 35 mm and 45 mm?</h3>

In [39]:
bill_range = penguins[(penguins['bill_length_mm'] >= 35) & (penguins['bill_length_mm'] <= 45)]

In [40]:
bill_range

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen        39.10000       18.70000         181.000000   
1       Adelie  Torgersen        39.50000       17.40000         186.000000   
2       Adelie  Torgersen        40.30000       18.00000         195.000000   
3       Adelie  Torgersen        43.92193       17.15117         200.915205   
4       Adelie  Torgersen        36.70000       19.30000         193.000000   
..         ...        ...             ...            ...                ...   
298  Chinstrap      Dream        43.20000       16.60000         187.000000   
306  Chinstrap      Dream        40.90000       16.60000         187.000000   
308  Chinstrap      Dream        42.50000       16.70000         187.000000   
330  Chinstrap      Dream        42.50000       17.30000         187.000000   
340  Chinstrap      Dream        43.50000       18.10000         202.000000   

     body_mass_g     sex  year  
0    3750.000000    male  2007  
1    3800.000000  female  2007  
2    3250.000000  female  2007  
3    4201.754386    male  2007  
4    3450.000000  female  2007  
..           ...     ...   ...  
298  2900.000000  female  2007  
306  3200.000000  female  2008  
308  3350.000000  female  2008  
330  3350.000000  female  2009  
340  3400.000000  female  2009  

[170 rows x 8 columns]

<h4>INSIGHT:</h4>
<h3>This range may indicate average adult penguins, avoiding extremes.</h3>

<h3>Question 2</h3>
<h3>What is the longest bill length recorded in the dataset, and which species has it?</h3>

In [41]:
max_bill_length = penguins['bill_length_mm'].max()

max_bill_row = penguins[penguins['bill_length_mm'] == max_bill_length]

max_bill_species = max_bill_row['species'].values[0]

In [42]:
max_bill_length

np.float64(59.6)

In [43]:
max_bill_species

'Gentoo'

In [44]:
max_bill_row

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
185  Gentoo  Biscoe            59.6           17.0              230.0   

     body_mass_g   sex  year  
185       6050.0  male  2007

<h4>INSIGHT:</h4>
<h3>Gentoo penguins have the longest bill length (59.6 mm), possibly due to feeding adaptations or it may be an outlier.</h3>

<h3>Question 3</h3>
<h3>Which Adelie penguins have a bill depth less than 17 mm or a body mass greater than or equal to 4000 grams?</h3>

In [45]:
adelie_filtered = penguins[(penguins['species'] == 'Adelie') & ((penguins['bill_depth_mm'] < 17) | (penguins['body_mass_g'] >= 4000))]

In [46]:
adelie_filtered

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
3    Adelie  Torgersen        43.92193       17.15117         200.915205   
7    Adelie  Torgersen        39.20000       19.60000         195.000000   
9    Adelie  Torgersen        42.00000       20.20000         190.000000   
14   Adelie  Torgersen        34.60000       21.10000         198.000000   
17   Adelie  Torgersen        42.50000       20.70000         197.000000   
19   Adelie  Torgersen        46.00000       21.50000         194.000000   
30   Adelie      Dream        39.50000       16.70000         178.000000   
35   Adelie      Dream        39.20000       21.10000         196.000000   
39   Adelie      Dream        39.80000       19.10000         184.000000   
43   Adelie      Dream        44.10000       19.70000         196.000000   
44   Adelie      Dream        37.00000       16.90000         185.000000   
45   Adelie      Dream        39.60000       18.80000         190.000000   
49   Adelie      Dream        42.30000       21.20000         191.000000   
51   Adelie     Biscoe        40.10000       18.90000         188.000000   
53   Adelie     Biscoe        42.00000       19.50000         200.000000   
58   Adelie     Biscoe        36.50000       16.60000         181.000000   
60   Adelie     Biscoe        35.70000       16.90000         185.000000   
61   Adelie     Biscoe        41.30000       21.10000         195.000000   
63   Adelie     Biscoe        41.10000       18.20000         192.000000   
66   Adelie     Biscoe        35.50000       16.20000         195.000000   
67   Adelie     Biscoe        41.10000       19.10000         188.000000   
68   Adelie  Torgersen        35.90000       16.60000         190.000000   
69   Adelie  Torgersen        41.80000       19.40000         198.000000   
73   Adelie  Torgersen        45.80000       18.90000         197.000000   
75   Adelie  Torgersen        42.80000       18.50000         195.000000   
76   Adelie  Torgersen        40.90000       16.80000         191.000000   
78   Adelie  Torgersen        36.20000       16.10000         187.000000   
79   Adelie  Torgersen        42.10000       19.10000         195.000000   
81   Adelie  Torgersen        42.90000       17.60000         196.000000   
83   Adelie  Torgersen        35.10000       19.40000         193.000000   
91   Adelie      Dream        41.10000       18.10000         205.000000   
93   Adelie      Dream        39.60000       18.10000         186.000000   
95   Adelie      Dream        40.80000       18.90000         208.000000   
97   Adelie      Dream        40.30000       18.50000         196.000000   
98   Adelie      Dream        33.10000       16.10000         178.000000   
99   Adelie      Dream        43.20000       18.50000         192.000000   
101  Adelie     Biscoe        41.00000       20.00000         203.000000   
102  Adelie     Biscoe        37.70000       16.00000         183.000000   
103  Adelie     Biscoe        37.80000       20.00000         190.000000   
109  Adelie     Biscoe        43.20000       19.00000         197.000000   
110  Adelie     Biscoe        38.10000       16.50000         198.000000   
111  Adelie     Biscoe        45.60000       20.30000         191.000000   
113  Adelie     Biscoe        42.20000       19.50000         197.000000   
115  Adelie     Biscoe        42.70000       18.30000         196.000000   
124  Adelie  Torgersen        35.20000       15.90000         186.000000   
125  Adelie  Torgersen        40.60000       19.00000         199.000000   
127  Adelie  Torgersen        41.50000       18.30000         195.000000   
129  Adelie  Torgersen        44.10000       18.00000         210.000000   
133  Adelie      Dream        37.50000       18.50000         199.000000   
138  Adelie      Dream        37.00000       16.50000         185.000000   
139  Adelie      Dream        39.70000       17.90000         193.000000   
142  Adelie      Dream        32.10000       15.5000

<h4>INSIGHT:</h4>
<h3>This mix of conditions captures both lighter-billed and heavier Adelies, showing diversity within the species.</h3>

<h3>Question 4</h3>
<h3>Which penguins have a body mass below the average mass and are from the year 2009?</h3>

In [47]:
avg_mass = penguins['body_mass_g'].mean()
below_avg_2009 = penguins[(penguins['body_mass_g'] < avg_mass) & (penguins['year'] == 2009)]

In [48]:
avg_mass

np.float64(4201.754385964912)

<h4>INSIGHT:</h4>
<h3>This could reveal environmental or food supply issues specific to that year affecting penguin health.</h3>

<h3>Question 5</h3>
<h3>How many penguins are not Gentoo species?</h3>

In [49]:
not_gentoo = penguins[penguins['species'] != 'Gentoo']

In [50]:
not_gentoo

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen        39.10000       18.70000         181.000000   
1       Adelie  Torgersen        39.50000       17.40000         186.000000   
2       Adelie  Torgersen        40.30000       18.00000         195.000000   
3       Adelie  Torgersen        43.92193       17.15117         200.915205   
4       Adelie  Torgersen        36.70000       19.30000         193.000000   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream        55.80000       19.80000         207.000000   
340  Chinstrap      Dream        43.50000       18.10000         202.000000   
341  Chinstrap      Dream        49.60000       18.20000         193.000000   
342  Chinstrap      Dream        50.80000       19.00000         210.000000   
343  Chinstrap      Dream        50.20000       18.70000         198.000000   

     body_mass_g     sex  year  
0    3750.000000    male  2007  
1    3800.000000  female  2007  
2    3250.000000  female  2007  
3    4201.754386    male  2007  
4    3450.000000  female  2007  
..           ...     ...   ...  
339  4000.000000    male  2009  
340  3400.000000  female  2009  
341  3775.000000    male  2009  
342  4100.000000    male  2009  
343  3775.000000  female  2009  

[220 rows x 8 columns]

<h4>INSIGHT:</h4>
<h3>This allows comparison between Gentoo and non-Gentoo species, useful for species-based analysis.</h3>

<h3>Question 6</h3>
<h3>Which penguins are not Chinstrap and have a flipper length greater than or equal to 200 mm?</h3>

In [51]:
not_chinstrap = penguins[(penguins['species'] != 'Chinstrap') & (penguins['flipper_length_mm'] >= 200)]

In [52]:
not_chinstrap

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
3    Adelie  Torgersen        43.92193       17.15117         200.915205   
53   Adelie     Biscoe        42.00000       19.50000         200.000000   
90   Adelie      Dream        35.70000       18.00000         202.000000   
91   Adelie      Dream        41.10000       18.10000         205.000000   
95   Adelie      Dream        40.80000       18.90000         208.000000   
..      ...        ...             ...            ...                ...   
271  Gentoo     Biscoe        43.92193       17.15117         200.915205   
272  Gentoo     Biscoe        46.80000       14.30000         215.000000   
273  Gentoo     Biscoe        50.40000       15.70000         222.000000   
274  Gentoo     Biscoe        45.20000       14.80000         212.000000   
275  Gentoo     Biscoe        49.90000       16.10000         213.000000   

     body_mass_g     sex  year  
3    4201.754386    male  2007  
53   4050.000000    male  2008  
90   3550.000000  female  2008  
91   4300.000000    male  2008  
95   4300.000000    male  2008  
..           ...     ...   ...  
271  4201.754386    male  2009  
272  4850.000000  female  2009  
273  5750.000000    male  2009  
274  5200.000000  female  2009  
275  5400.000000    male  2009  

[134 rows x 8 columns]

<h4>INSIGHT:</h4>
<h3>This focuses on large-flippered penguins from species other than Chinstrap, showing physical diversity.</h3>

<h3>Question 7</h3>
<h3>How many penguins have a bill depth greater than 18 mm or a flipper length greater than or equal to 210 mm?</h3>

In [53]:
filter_penguins = penguins[(penguins['bill_depth_mm'] > 18) | (penguins['flipper_length_mm'] >= 210)]

In [54]:
filter_penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
4       Adelie  Torgersen            36.7           19.3              193.0   
5       Adelie  Torgersen            39.3           20.6              190.0   
7       Adelie  Torgersen            39.2           19.6              195.0   
8       Adelie  Torgersen            34.1           18.1              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
4         3450.0  female  2007  
5         3650.0    male  2007  
7         4675.0    male  2007  
8         3475.0    male  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[240 rows x 8 columns]

<h4>INSIGHT:</h4>
<h3>This captures penguins with at least one large physical feature, which could relate to strength or feeding ability.</h3>

<h3>How to create a column</h3>

In [55]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    category
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      344 non-null    float64 
 4   flipper_length_mm  344 non-null    float64 
 5   body_mass_g        344 non-null    float64 
 6   sex                344 non-null    category
 7   year               344 non-null    category
dtypes: category(4), float64(4)
memory usage: 12.7 KB


In [56]:
penguins.columns

Index(['species', 'island', 'bill_length_mm', 'bill_depth_mm',
       'flipper_length_mm', 'body_mass_g', 'sex', 'year'],
      dtype='object')

In [57]:
penguins['mass_kg'] = round (penguins['body_mass_g']/1000,1)

In [58]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen        39.10000       18.70000         181.000000   
1       Adelie  Torgersen        39.50000       17.40000         186.000000   
2       Adelie  Torgersen        40.30000       18.00000         195.000000   
3       Adelie  Torgersen        43.92193       17.15117         200.915205   
4       Adelie  Torgersen        36.70000       19.30000         193.000000   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream        55.80000       19.80000         207.000000   
340  Chinstrap      Dream        43.50000       18.10000         202.000000   
341  Chinstrap      Dream        49.60000       18.20000         193.000000   
342  Chinstrap      Dream        50.80000       19.00000         210.000000   
343  Chinstrap      Dream        50.20000       18.70000         198.000000   

     body_mass_g     sex  year  mass_kg  
0    3750.000000    male  2007      3.8  
1    3800.000000  female  2007      3.8  
2    3250.000000  female  2007      3.2  
3    4201.754386    male  2007      4.2  
4    3450.000000  female  2007      3.4  
..           ...     ...   ...      ...  
339  4000.000000    male  2009      4.0  
340  3400.000000  female  2009      3.4  
341  3775.000000    male  2009      3.8  
342  4100.000000    male  2009      4.1  
343  3775.000000  female  2009      3.8  

[344 rows x 9 columns]

In [59]:
penguins['mass_kg'].max()

np.float64(6.3)

In [60]:
penguins[penguins['mass_kg'] == 6.3]

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
169  Gentoo  Biscoe            49.2           15.2              221.0   

     body_mass_g   sex  year  mass_kg  
169       6300.0  male  2007      6.3

<h3>Question 8</h3>
<h3>
Do the three penguin species (Adelie, Chinstrap, Gentoo) significantly differ in terms of body mass?</h3>

In [61]:
from scipy.stats import f_oneway

In [62]:
stats, p_value = f_oneway(
    penguins[penguins['species'] == 'Adelie']['body_mass_g'],
    penguins[penguins['species'] == 'Chinstrap']['body_mass_g'],
    penguins[penguins['species'] == 'Gentoo']['body_mass_g']
)


In [63]:
p_value

np.float64(1.0069770504490732e-81)

<h4>INSIGHT:</h4>
<h3>Body mass varies by specie, Gentoo are the heaviest, Adelie the lightest.</h3>

<h3>Question 9
</h3>
<h3>Is there a significant difference in bill depth between male and female penguins?</h3>

In [64]:
stats, p_value = f_oneway(
    penguins[penguins['sex'] == 'male']['bill_depth_mm'],
    penguins[penguins['sex'] == 'female']['bill_depth_mm']
)


In [65]:
p_value

np.float64(1.2906481751545917e-11)

<h4>INSIGHT:</h4>
<h3>The difference in bill depth is smaller, but still may be statistically significant</h3>

<h3>Question 10</h3>
<h3>
Among male penguins, is there a significant difference in body mass across the three species?</h3>

In [66]:
stats, p_value = f_oneway(
    penguins[(penguins['species'] == 'Adelie') & (penguins['sex'] == 'male')]['body_mass_g'],
    penguins[(penguins['species'] == 'Chinstrap') & (penguins['sex'] == 'male')]['body_mass_g'],
    penguins[(penguins['species'] == 'Gentoo') & (penguins['sex'] == 'male')]['body_mass_g']
)

In [67]:
p_value

np.float64(4.655047128307857e-56)

<h4>INSIGHT:</h4>
<h3>The test shows whether species has an effect on male penguins body mass.</h3>

<h3>Question 11</h3>
<h3>
Within Chinstrap penguins, is there a significant difference in bill depth between males and females?</h3>

In [68]:
stats, p_value = f_oneway(
    penguins[(penguins['species'] == 'Chinstrap') & (penguins['sex'] == 'male')]['bill_depth_mm'],
    penguins[(penguins['species'] == 'Chinstrap') & (penguins['sex'] == 'female')]['bill_depth_mm']
)


In [69]:
p_value

np.float64(6.563879282777359e-13)

<h4>INSIGHT:</h4>
<h3>Male Chinstrap penguins typically have slightly deeper bills than females.</h3>

<h3>Question 12</h3>
<h3>
Do male penguins across Adelie, Chinstrap, and Gentoo species significantly differ in flipper length?</h3>

In [74]:
stats, p_value = f_oneway(
    penguins[(penguins['species'] == 'Adelie') & (penguins['sex'] == 'male')]['flipper_length_mm'],
    penguins[(penguins['species'] == 'Chinstrap') & (penguins['sex'] == 'male')]['flipper_length_mm'],
    penguins[(penguins['species'] == 'Gentoo') & (penguins['sex'] == 'male')]['flipper_length_mm']
)

In [75]:
p_value

np.float64(5.824007653367733e-64)

<h4>INSIGHT:</h4>
<h3>Male Gentoo penguins are expected to have the longest flippers.</h3>

<h3>Question 13</h3>
<h3>
Do male and female penguins significantly differ in terms of bill length?</h3>

In [78]:
stats, p_value = f_oneway(
    penguins[penguins['sex'] == 'male']['bill_length_mm'],
    penguins[penguins['sex'] == 'female']['bill_length_mm']
)

In [79]:
p_value

np.float64(9.307501289283404e-10)

<h4>INSIGHT:</h4>
<h3>Male penguins generally have longer bills, p-value < 0.05 expected.</h3>